osquery/tables/networking/windows/defender_firewall_rules.cpp

```diff
@@ -0,0 +1,191 @@
+  if (FAILED(hr)) {
+    TLOG << "Failed to get firewall rules";
+    return results;
+  }
+
+  CComPtr<IUnknown> enumerator;
+  hr = rules->get__NewEnum(&enumerator);
+  if (FAILED(hr)) {
+    TLOG << "Failed to get firewall rules enumerator";
+    return results;
+  }
+
+  CComPtr<IEnumVARIANT> enumvar;
+  hr = enumerator->QueryInterface(IID_PPV_ARGS(&enumvar));
+  if (FAILED(hr)) {
+    TLOG << "Failed to get firewall rules enumerator interface";
+    return results;
+  }
+
+  do {
+    CComVariant value;
+    ULONG fetched = 0;
+    hr = enumvar->Next(1, &value, &fetched);
+    if (FAILED(hr)) {
+      TLOG << "Failed to enumerate next firewall rule";
+      break;
+    }
+
+    if (hr != S_FALSE) {
+      hr = value.ChangeType(VT_DISPATCH);
+      if (FAILED(hr)) {
+        TLOG << "Failed to get IDispatch for enumerated rule";
+        break;
+      }
+
+      CComPtr<INetFwRule> rule;
+      hr = (V_DISPATCH(&value))
+               ->QueryInterface(__uuidof(INetFwRule), (void**)&rule);
+      if (FAILED(hr)) {
+        TLOG << "Failed to get firewall rule interface";
+        break;
+      }
+
+      Row r;
+      hr = populateRow(rule, r);
+      if (FAILED(hr)) {
+        TLOG << "Failed to convert firewall rule to row";
+        break;
+      }
+
+      results.push_back(std::move(r));
+    }
+  } while (SUCCEEDED(hr) && hr != S_FALSE);
+
+  return results;
+}
+
+HRESULT populateRow(INetFwRule* rule, Row& r) {
+  HRESULT hr = S_OK;
+
+  CComBSTR name;
+  if (FAILED(hr = rule->get_Name(&name))) {
+    return hr;
+  }
+  r["name"] = bstrToString(name);
+
+  CComBSTR appname;
+  if (FAILED(hr = rule->get_ApplicationName(&appname))) {
+    return hr;
+  }
+  r["app_name"] = bstrToString(appname);
+
+  NET_FW_ACTION action;
+  if (FAILED(hr = rule->get_Action(&action))) {
+    return hr;
+  }
+  switch (action) {
+  case NET_FW_ACTION_BLOCK:
+    r["action"] = "Block";
+    break;
+  case NET_FW_ACTION_ALLOW:
+    r["action"] = "Allow";
+    break;
+  }
+
+  VARIANT_BOOL enabled;
+  if (FAILED(hr = rule->get_Enabled(&enabled))) {
+    return hr;
+  }
+  r["enabled"] = INTEGER(bool(enabled));
+
+  NET_FW_RULE_DIRECTION direction;
+  if (FAILED(hr = rule->get_Direction(&direction))) {
+    return hr;
+  }
+  switch (direction) {
+  case NET_FW_RULE_DIR_IN:
+    r["direction"] = "In";
+    break;
+  case NET_FW_RULE_DIR_OUT:
+    r["direction"] = "Out";
+    break;
+  }
+
+  long protocol = 0;
+  if (FAILED(hr = rule->get_Protocol(&protocol))) {
+    return hr;
+  }
+  switch (protocol) {
+  case NET_FW_IP_PROTOCOL_TCP:
+    r["protocol"] = "TCP";
+    break;
+  case NET_FW_IP_PROTOCOL_UDP:
+    r["protocol"] = "UDP";
+    break;
+  case NET_FW_IP_PROTOCOL_ANY:
+    r["protocol"] = "Any";
+    break;
+  }
+
+  long profileBitmask = 0;
+  if (FAILED(hr = rule->get_Profiles(&profileBitmask))) {
+    return hr;
+  }
+
+  std::string profiles;
+  if (profileBitmask == NET_FW_PROFILE2_ALL) {
+    r["profiles"] = "Any";
+  } else {
```

Comment:
> * Renamed `profile` into `profiles` column that contains decoded comma separated text values like `Private,Public`.  Could create individual rows for each if needed. Let me know.

I think this pattern is hard to work with -- It is quite difficult to filter based on these. I would probably make them into boolean columns. Not sure if `any` should be a column, vs just setting true on the others. (Exposing a bitmask would also work, but isn't very friendly)

Reply:
yeah, was not sure about this one. sure, can split into three boolean columns:
```profile_domain```, ```profile_private```, ```profile_public```

according to the enum
```
typedef 
enum NET_FW_PROFILE_TYPE2_
    {
        NET_FW_PROFILE2_DOMAIN	= 0x1,
        NET_FW_PROFILE2_PRIVATE	= 0x2,
        NET_FW_PROFILE2_PUBLIC	= 0x4,
        NET_FW_PROFILE2_ALL	= 0x7fffffff
    } 	NET_FW_PROFILE_TYPE2;
```

https://docs.microsoft.com/en-us/windows/win32/api/icftypes/ne-icftypes-net_fw_profile_type2

Reply:
something like this:
<img width="1986" alt="Screen Shot 2021-12-01 at 11 18 47 AM" src="https://user-images.githubusercontent.com/872351/144271758-be96c868-fdf3-4deb-89ca-5eb0addc6758.png">